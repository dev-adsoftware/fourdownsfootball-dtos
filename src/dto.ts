import { ClassConstructor, plainToClass } from 'class-transformer';
import { validateSync, ValidationError } from 'class-validator';

export abstract class Dto {
  public init(obj: Record<string | number, unknown>): this {
    Object.assign(
      this,
      plainToClass(this.constructor as ClassConstructor<this>, { ...obj }),
    );
    return this.validate(true);
  }

  public serialize(): string {
    return JSON.stringify(this);
  }

  private validate(whitelist: boolean): this {
    const errors = validateSync(this, {
      whitelist,
      validationError: { target: false },
    });
    if (errors.length > 0) {
      throw new Error(
        `${this.constructor.name}: ${this.validationErrors(errors)}`,
      );
    }

    return this;
  }

  private validationErrors(errors: ValidationError[]): string {
    return errors
      .map((e) => {
        if (e.constraints) return Object.values(e.constraints).join(', ');
        return String(e);
      })
      .join(', ');
  }
}
