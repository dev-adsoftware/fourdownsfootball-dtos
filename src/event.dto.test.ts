/* eslint-disable max-classes-per-file */
import { createDto } from './dto';
import { Event } from '.';

describe('given: no event data', () => {
  describe('when: I create an event dto', () => {
    it('then: dto threw expected error', async () => {
      expect(() => createDto(Event, {})).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('given: proper event', () => {
  describe('when: I create an event dto', () => {
    it('then: dto was created with expected values', async () => {
      expect(
        createDto(Event, {
          id: 'jest-id',
          date: 'jest-date',
          source: 'jest-source',
          type: 'jest-type',
          version: 'jest-version',
          data: 'jest-data',
        }),
      ).toEqual({
        data: 'jest-data',
        date: 'jest-date',
        id: 'jest-id',
        source: 'jest-source',
        type: 'jest-type',
        version: 'jest-version',
      });
    });
  });
});
