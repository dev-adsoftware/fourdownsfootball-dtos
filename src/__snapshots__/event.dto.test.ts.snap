// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given: no event data when: I create an event dto then: dto threw expected error 1`] = `"Event: id must be a string, date must be a string, source must be a string, type must be a string, version must be a string, data must be a string"`;
